In [2]:
from pathlib import Path

import duckdb
import pandas as pd
from tabulate import tabulate

data_root = Path("D:\\데이터\\건축데이터 건축허브 개방데이터")
path_DB = data_root / "건축물대장_2025년_02월.db"

건축허브에서 제공하는 건축물대장 데이터를 활용하여 분석

2025년 2월 이전 데이터는 위반건축물 등 일부 건축물이 빠져있는 문제가 있어 부득이 2025년 2월 말 데이터로 2024년 말 기준 건축물 현황을 집계

In [3]:
# Open a DuckDB connection
con = duckdb.connect(database=path_DB, read_only=True)

# print the list of tables in the database
tables = con.execute("SHOW TABLES").fetchall()
print("Tables in the database:")
for table in tables:
    print(table[0])
    print()

# show heads of the tables
for table in tables:
    table_name = table[0]
    print(f"Head of {table_name}:")
    df = con.execute(f"SELECT * FROM {table_name} LIMIT 5").fetchdf()
    print(df)
    print()

Tables in the database:
기본개요

총괄표제부

층별개요

표제부

Head of 기본개요:
  관리_건축물대장_PK 관리_상위_건축물대장_PK 대장_구분_코드 대장_구분_코드_명 대장_종류_코드 대장_종류_코드_명  \
0      100211           None        1         일반        1      총괄표제부   
1      100215           None        1         일반        1      총괄표제부   
2      100216           None        1         일반        1      총괄표제부   
3      100218           None        1         일반        1      총괄표제부   
4      100219           None        1         일반        1      총괄표제부   

                  대지_위치                  도로명_대지_위치  건물_명 시군구_코드  ...  \
0  서울특별시 종로구 관철동 12-1번지   서울특별시 종로구 종로14길 20 (관철동)  None  11110  ...   
1  서울특별시 종로구 청진동 21-1번지  서울특별시 종로구 종로5길 32-6 (청진동)  None  11110  ...   
2  서울특별시 종로구 중학동 16-2번지                       None  None  11110  ...   
3  서울특별시 종로구 중학동 24-1번지      서울특별시 종로구 율곡로 5 (중학동)  None  11110  ...   
4  서울특별시 종로구 중학동 35-1번지                       None  None  11110  ...   

  새주소_지상지하_코드 새주소_본_번 새주소_부_번 지역_코드 지구_코드 구역_코드 지역_코드_명  지구_코드_명 구역_코드_명

In [4]:
# Get the list of tables in the database
tables = con.execute("SHOW TABLES").fetchall()

print("=== COLUMN COUNT SUMMARY ===")
column_counts = []

for table in tables:
    table_name = table[0]
    columns_info = con.execute(f"DESCRIBE {table_name}").fetchdf()
    column_count = len(columns_info)
    column_names = list(columns_info["column_name"])

    column_counts.append(
        {
            "Table Name": table_name,
            "Column Count": column_count,
            "Column Names": ", ".join(column_names),
        }
    )

column_counts_df = pd.DataFrame(column_counts)
# Use context manager to temporarily set pandas display options
with pd.option_context(
    "display.max_columns",
    None,
    "display.max_colwidth",
    None,
    "display.width",
    None,
    "display.max_rows",
    None,
):
    display(column_counts_df)

=== COLUMN COUNT SUMMARY ===


,Table Name,Column Count,Column Names
0,기본개요,30,"관리_건축물대장_PK, 관리_상위_건축물대장_PK, 대장_구분_코드, 대장_구분_코드_명, 대장_종류_코드, 대장_종류_코드_명, 대지_위치, 도로명_대지_위치, 건물_명, 시군구_코드, 법정동_코드, 대지_구분_코드, 번, 지, 특수지_명, 블록, 로트, 외필지_수, 새주소_도로_코드, 새주소_법정동_코드, 새주소_지상지하_코드, 새주소_본_번, 새주소_부_번, 지역_코드, 지구_코드, 구역_코드, 지역_코드_명, 지구_코드_명, 구역_코드_명, 생성_일자"
1,총괄표제부,64,"관리_건축물대장_PK, 대장_구분_코드, 대장_구분_코드_명, 대장_종류_코드, 대장_종류_코드_명, 신_구_대장_구분_코드, 신_구_대장_구분_코드_명, 대지_위치, 도로명_대지_위치, 건물_명, 시군구_코드, 법정동_코드, 대지_구분_코드, 번, 지, 특수지_명, 블록, 로트, 외필지_수, 새주소_도로_코드, 새주소_법정동_코드, 새주소_지상지하_코드, 새주소_본_번, 새주소_부_번, 대지_면적(㎡), 건축_면적(㎡), 건폐_율(%), 연면적(㎡), 용적_률_산정_연면적(㎡), 용적_률(%), 주_용도_코드, 주_용도_코드_명, 기타_용도, 세대_수(세대), 가구_수(가구), 주_건축물_수, 부속_건축물_수, 부속_건축물_면적(㎡), 총_주차_수, 옥내_기계식_대수(대), 옥내_기계식_면적(㎡), 옥외_기계식_대수(대), 옥외_기계식_면적(㎡), 옥내_자주식_대수(대), 옥내_자주식_면적(㎡), 옥외_자주식_대수(대), 옥외_자주식_면적(㎡), 허가_일, 착공_일, 사용승인_일, 허가번호_년, 허가번호_기관_코드, 허가번호_기관_코드_명, 허가번호_구분_코드, 허가번호_구분_코드_명, 호_수(호), 에너지효율_등급, 에너지절감_율, 에너지_EPI점수, 친환경_건축물_등급, 친환경_건축물_인증점수, 지능형_건축물_등급, 지능형_건축물_인증점수, 생성_일자"
2,층별개요,33,"관리_건축물대장_PK, 대지_위치, 도로명_대지_위치, 건물_명, 시군구_코드, 법정동_코드, 대지_구분_코드, 번, 지, 특수지_명, 블록, 로트, 새주소_도로_코드, 새주소_법정동_코드, 새주소_지상지하_코드, 새주소_본_번, 새주소_부_번, 동_명, 층_구분_코드, 층_구분_코드_명, 층_번호, 층_번호_명, 구조_코드, 구조_코드_명, 기타_구조, 주_용도_코드, 주_용도_코드_명, 기타_용도, 면적(㎡), 주_부속_구분_코드, 주_부속_구분_코드_명, 면적_제외_여부, 생성_일자"
3,표제부,77,"관리_건축물대장_PK, 대장_구분_코드, 대장_구분_코드_명, 대장_종류_코드, 대장_종류_코드_명, 대지_위치, 도로명_대지_위치, 건물_명, 시군구_코드, 법정동_코드, 대지_구분_코드, 번, 지, 특수지_명, 블록, 로트, 외필지_수, 새주소_도로_코드, 새주소_법정동_코드, 새주소_지상지하_코드, 새주소_본_번, 새주소_부_번, 동_명, 주_부속_구분_코드, 주_부속_구분_코드_명, 대지_면적(㎡), 건축_면적(㎡), 건폐_율(%), 연면적(㎡), 용적_률_산정_연면적(㎡), 용적_률(%), 구조_코드, 구조_코드_명, 기타_구조, 주_용도_코드, 주_용도_코드_명, 기타_용도, 지붕_코드, 지붕_코드_명, 기타_지붕, 세대_수(세대), 가구_수(가구), 높이(m), 지상_층_수, 지하_층_수, 승용_승강기_수, 비상용_승강기_수, 부속_건축물_수, 부속_건축물_면적(㎡), 총_동_연면적(㎡), 옥내_기계식_대수(대), 옥내_기계식_면적(㎡), 옥외_기계식_대수(대), 옥외_기계식_면적(㎡), 옥내_자주식_대수(대), 옥내_자주식_면적(㎡), 옥외_자주식_대수(대), 옥외_자주식_면적(㎡), 허가_일, 착공_일, 사용승인_일, 허가번호_년, 허가번호_기관_코드, 허가번호_기관_코드_명, 허가번호_구분_코드, 허가번호_구분_코드_명, 호_수(호), 에너지효율_등급, 에너지절감_율, 에너지_EPI점수, 친환경_건축물_등급, 친환경_건축물_인증점수, 지능형_건축물_등급, 지능형_건축물_인증점수, 생성_일자, 내진_설계_적용_여부, 내진_능력"


In [5]:
# Create a list to store the table names and their record counts
table_counts = []

# Iterate through the tables and count the records
for table in tables:
    table_name = table[0]
    count = con.execute(f"SELECT COUNT(*) FROM {table_name}").fetchone()[0]
    table_counts.append({"Table Name": table_name, "Record Count": count})

# Convert the list to a DataFrame
record_counts_df = pd.DataFrame(table_counts)

# Display the DataFrame
display(record_counts_df)

,Table Name,Record Count
0,기본개요,27994157
1,총괄표제부,615541
2,층별개요,21048715
3,표제부,8027067


In [6]:
table_name = "표제부"


def get_column_counts_duckdb(con: duckdb.DuckDBPyConnection, table_name: str):
    # 0) print the statement
    print(f"Counting columns in table: {table_name}")
    # 1) 테이블 스키마에서 컬럼 이름을 가져오고
    cols_df = con.execute(f"PRAGMA table_info('{table_name}')").fetchdf()
    cols = cols_df["name"].tolist()

    # 2) 각 컬럼에 대한 COUNT 문을 생성
    # 표제부에는 컬럼명에 %가 들어간 경우가 있어, 오류가 발생함
    # 이를 피하기 위해, 컬럼명을 따옴표로 감싸줌
    count_exprs = [f'COUNT("{col}") AS "{table_name}_{col}_count"' for col in cols]

    # 3) 최종 쿼리 조립
    query = """
    SELECT
        {exprs}
    FROM "{table_name}"
    """.format(exprs=",\n    ".join(count_exprs), table_name=table_name)

    # 4) 실행 및 결과 반환
    result = con.execute(query).fetchdf().T

    return result


with pd.option_context("display.max_rows", None, "display.max_columns", None):
    # Display the result with all rows and columns
    print(
        tabulate(get_column_counts_duckdb(con, table_name), intfmt=",", floatfmt=",.0f")
    )

Counting columns in table: 표제부
------------------------------------  ---------
표제부_관리_건축물대장_PK_count       8,027,067
표제부_대장_구분_코드_count           8,027,067
표제부_대장_구분_코드_명_count        8,027,067
표제부_대장_종류_코드_count           8,027,067
표제부_대장_종류_코드_명_count        8,027,067
표제부_대지_위치_count                8,027,067
표제부_도로명_대지_위치_count         7,076,491
표제부_건물_명_count                  1,160,843
표제부_시군구_코드_count              8,027,067
표제부_법정동_코드_count              8,027,063
표제부_대지_구분_코드_count           8,027,059
표제부_번_count                       8,027,067
표제부_지_count                       8,027,055
표제부_특수지_명_count                   33,202
표제부_블록_count                        12,772
표제부_로트_count                         8,734
표제부_외필지_수_count                8,027,067
표제부_새주소_도로_코드_count         7,076,491
표제부_새주소_법정동_코드_count       7,074,579
표제부_새주소_지상지하_코드_count     8,024,683
표제부_새주소_본_번_count             7,174,022
표제부_새주소_부_번_count             5,375,334
표제부_동_명_count                    2,199,542

In [7]:
table_name = "표제부"
query = f"""SELECT
    *
FROM
    (SUMMARIZE "{table_name}")
WHERE
    column_name IN (
        '시군구_코드',
        '법정동_코드',
        '주_부속_구분_코드',
        '주_부속_구분_코드_명',
        '대지_면적(㎡)',
        '건축_면적(㎡)',
        '건폐_율(%)',
        '연면적(㎡)',
        '용적_률_산정_연면적(㎡)',
        '용적_률(%)',
        '구조_코드',
        '구조_코드_명',
        '기타_구조',
        '주_용도_코드',
        '주_용도_코드_명',
        '기타_용도',
        '높이(m)',
        '지상_층_수',
        '지하_층_수',
        '부속_건축물_수',
        '부속_건축물_면적(㎡)',
        '총_동_연면적(㎡)',
        '허가_일',
        '착공_일',
        '사용승인_일',
    );
"""
con.sql(query).show(max_rows=1_000)

┌─────────────────────────┬───────────────┬─────────────────┬────────────────┬───────────────┬─────────────────────┬────────────────────┬─────────┬─────────┬─────────┬─────────┬─────────────────┐
│       column_name       │  column_type  │       min       │      max       │ approx_unique │         avg         │        std         │   q25   │   q50   │   q75   │  count  │ null_percentage │
│         varchar         │    varchar    │     varchar     │    varchar     │     int64     │       varchar       │      varchar       │ varchar │ varchar │ varchar │  int64  │  decimal(9,2)   │
├─────────────────────────┼───────────────┼─────────────────┼────────────────┼───────────────┼─────────────────────┼────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────────────┤
│ 시군구_코드             │ VARCHAR       │ 0               │ 99999          │           252 │ NULL                │ NULL               │ NULL    │ NULL    │ NULL    │ 8027067 │            0.00 │
│ 법정동_코드             │ VA

표제부 면적 관련 데이터는 미기재율 1% 미만.
그러나 면적, 건폐율, 용적률 등이 음수인 경우, 이론상 100%를 초과할 수 없는 건폐율이 1,000,000,000인 경우 등 이상값 발견됨.

시군구, 법정동, 사용승인일 등은 미기재율은 낮으나 이상값이 발견됨 (0, 99999, 9990408 등)

In [8]:
with pd.option_context("display.max_rows", None, "display.max_columns", None):
    # Display the result with all rows and columns
    print(
        tabulate(
            get_column_counts_duckdb(con, "총괄표제부"), intfmt=",", floatfmt=",.0f"
        )
    )

Counting columns in table: 총괄표제부
----------------------------------------  -------
총괄표제부_관리_건축물대장_PK_count       615,541
총괄표제부_대장_구분_코드_count           615,541
총괄표제부_대장_구분_코드_명_count        615,487
총괄표제부_대장_종류_코드_count           615,541
총괄표제부_대장_종류_코드_명_count        615,541
총괄표제부_신_구_대장_구분_코드_count     613,920
총괄표제부_신_구_대장_구분_코드_명_count  613,920
총괄표제부_대지_위치_count                615,541
총괄표제부_도로명_대지_위치_count         385,698
총괄표제부_건물_명_count                  101,028
총괄표제부_시군구_코드_count              615,540
총괄표제부_법정동_코드_count              615,540
총괄표제부_대지_구분_코드_count           615,540
총괄표제부_번_count                       615,540
총괄표제부_지_count                       615,534
총괄표제부_특수지_명_count                  1,479
총괄표제부_블록_count                         993
총괄표제부_로트_count                         690
총괄표제부_외필지_수_count                615,541
총괄표제부_새주소_도로_코드_count         385,698
총괄표제부_새주소_법정동_코드_count       385,407
총괄표제부_새주소_지상지하_코드_count     609,222
총괄표제부_새주소_본_번_count             417,411
총괄표제부

In [9]:
table_name = "총괄표제부"
query = f"""SELECT
    *
FROM
    (SUMMARIZE "{table_name}")
WHERE
    column_name IN (
        '시군구_코드',
        '법정동_코드',
        '주_부속_구분_코드',
        '주_부속_구분_코드_명',
        '대지_면적(㎡)',
        '건축_면적(㎡)',
        '건폐_율(%)',
        '연면적(㎡)',
        '용적_률_산정_연면적(㎡)',
        '용적_률(%)',
        '주_용도_코드',
        '주_용도_코드_명',
        '기타_용도',
        '세대_수(세대)',
        '가구_수(가구)',
        '주_건축물_수',
        '부속_건축물_수',
        '부속_건축물_면적(㎡)',
        '허가_일',
        '착공_일',
        '사용승인_일',
    );
"""
con.sql(query).show(max_rows=1_000)

┌─────────────────────────┬───────────────┬────────────┬─────────────────────────────┬───────────────┬────────────────────┬────────────────────┬─────────┬─────────┬─────────┬────────┬─────────────────┐
│       column_name       │  column_type  │    min     │             max             │ approx_unique │        avg         │        std         │   q25   │   q50   │   q75   │ count  │ null_percentage │
│         varchar         │    varchar    │  varchar   │           varchar           │     int64     │      varchar       │      varchar       │ varchar │ varchar │ varchar │ int64  │  decimal(9,2)   │
├─────────────────────────┼───────────────┼────────────┼─────────────────────────────┼───────────────┼────────────────────┼────────────────────┼─────────┼─────────┼─────────┼────────┼─────────────────┤
│ 시군구_코드             │ VARCHAR       │ 11110      │ 99999                       │           252 │ NULL               │ NULL               │ NULL    │ NULL    │ NULL    │ 615541 │            0.

In [10]:
query = """
SELECT 
    m.관리_건축물대장_PK, 
    m.관리_상위_건축물대장_PK,
    m.관리_상위_건축물대장_PK IS NULL AS 관리_상위_건축물대장_PK_null,
    m.대지_위치,
    m.건물_명,
    b.동_명,
    b."시군구_코드",
    b."법정동_코드",
    b."주_부속_구분_코드",
    b."주_부속_구분_코드_명",
    b."대지_면적(㎡)",
    b."건축_면적(㎡)",
    b."건폐_율(%)",
    b."연면적(㎡)",
    b."용적_률_산정_연면적(㎡)",
    b."용적_률(%)",
    b."주_용도_코드",
    b."주_용도_코드_명",
    b."기타_용도",
    b."높이(m)",
    b."지상_층_수",
    b."지하_층_수",
    b."부속_건축물_수",
    b."부속_건축물_면적(㎡)",
    b."총_동_연면적(㎡)",
    b."허가_일",
    b."착공_일",
    b."사용승인_일",
    b.생성_일자,
    r."대지_면적(㎡)"               AS "총괄_대지_면적(㎡)",
    r."건축_면적(㎡)"               AS "총괄_건축_면적(㎡)",
    r."건폐_율(%)"                  AS "총괄_건폐_율(%)",
    r."연면적(㎡)"                  AS "총괄_연면적(㎡)",
    r."용적_률_산정_연면적(㎡)"     AS "총괄_용적_률_산정_연면적(㎡)",
    r."용적_률(%)"                  AS "총괄_용적_률(%)",
    r."주_용도_코드"                AS "총괄_주_용도_코드",
    r."주_용도_코드_명"             AS "총괄_주_용도_코드_명",
    r."기타_용도"                   AS "총괄_기타_용도",
    r."세대_수(세대)"               AS "총괄_세대_수(세대)",
    r."가구_수(가구)"               AS "총괄_가구_수(가구)",
    r."주_건축물_수"                AS "총괄_주_건축물_수",
    r."부속_건축물_수"              AS "총괄_부속_건축물_수",
    r."부속_건축물_면적(㎡)"        AS "총괄_부속_건축물_면적(㎡)",
FROM 기본개요 m
INNER JOIN 표제부 b
ON m.관리_건축물대장_PK = b.관리_건축물대장_PK
LEFT JOIN 총괄표제부 r  -- short for "recap"
ON m.관리_상위_건축물대장_PK = r.관리_건축물대장_PK
"""
기본개요_표제부_총괄표제부 = con.sql(query)
기본개요_표제부_총괄표제부.limit(5).show()

┌────────────────────┬─────────────────────────┬──────────────────────────────┬──────────────────────────────────┬──────────────┬─────────────────┬─────────────┬─────────────┬───────────────────┬──────────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬──────────────┬─────────────────┬─────────────────────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬──────────┬──────────┬─────────────┬───────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬──────────────────────┬────────────────────────┬────────────────────┬────────────────────┬───────────────────┬─────────────────────┬───────────────────────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │ 관리_상위_건축물대장_PK_null │            대지_위치             │   건물_명    │      동_명      │ 시군구_코드 │ 법정동_코드 │ 주_부속_구분_코드 │ 주_부속_구분_코드_명 │ 

In [45]:
con.sql("""
        SELECT * 
        FROM 기본개요_표제부_총괄표제부
        WHERE 대지_위치 LIKE '서울특별시 강남구 대치동 500번지' -- AND 동_명 LIKE ''
        """).show()

┌────────────────────┬─────────────────────────┬──────────────────────────────┬──────────────────────────────────┬───────────────────────┬─────────┬─────────────┬─────────────┬───────────────────┬──────────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬──────────────┬─────────────────┬───────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬──────────┬─────────┬─────────────┬───────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬──────────────────────┬────────────────┬────────────────────┬────────────────────┬───────────────────┬─────────────────────┬───────────────────────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │ 관리_상위_건축물대장_PK_null │            대지_위치             │        건물_명        │  동_명  │ 시군구_코드 │ 법정동_코드 │ 주_부속_구분_코드 │ 주_부속_구분_코드_명 │ 대지_면적(㎡) │ 건축_면적(㎡) │

대치동 500번지 개포우성2차 단지는 총괄표제부에 대지면적 0으로 기재되어, 건폐율, 용적률도 0으로 표시됨.
(토지이용계획 상 실제 면적 36,959.8 ㎡)

표제부 대지면적, 건폐율, 용적률도 미기재(빈 칸)가 아니라 0으로 기재되어 있는데, 개방데이터 상에서는 동일하게 0으로 구분할 수 없음.

In [12]:
con.sql("""
        SELECT * 
        FROM 기본개요_표제부_총괄표제부
        WHERE 대지_위치 LIKE '%한솔동 939%' AND 동_명 LIKE '101동'
        """).show()

┌────────────────────┬─────────────────────────┬──────────────────────────────┬───────────────────────────────┬──────────────┬─────────┬─────────────┬─────────────┬───────────────────┬──────────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬──────────────┬─────────────────┬──────────────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬──────────┬──────────┬─────────────┬───────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬──────────────────────┬──────────────────┬────────────────────┬────────────────────┬───────────────────┬─────────────────────┬───────────────────────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │ 관리_상위_건축물대장_PK_null │           대지_위치           │   건물_명    │  동_명  │ 시군구_코드 │ 법정동_코드 │ 주_부속_구분_코드 │ 주_부속_구분_코드_명 │ 대지_면적(㎡) │ 건축_면적(㎡) │  건폐_율(%)   │ 

한솔동 939번지, 101동: 표제부 대장상에서 대지면적, 건폐율, 용적률이 0으로 기재가 아니라 미기재(빈 칸)된 경우.
개방데이터 상에서는 동일하게 0으로 구분할 수 없음.

In [13]:
con.sql("SELECT COUNT(*) FROM 기본개요_표제부_총괄표제부").show()

┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│      8027067 │
└──────────────┘



In [14]:
query = """
SELECT
  관리_상위_건축물대장_PK_null,
  주_부속_구분_코드,
  COUNT(*) AS 건축물_수,
  AVG("대지_면적(㎡)") AS 평균_대지_면적,
  AVG("건축_면적(㎡)") AS 평균_건축_면적,
  AVG("건폐_율(%)") AS 평균_건폐율,
  AVG("연면적(㎡)") AS 평균_연면적,
  AVG("용적_률_산정_연면적(㎡)") AS 평균_용적률산정용연면적,
  AVG("용적_률(%)") AS 평균_용적률,
  AVG(지상_층_수) AS 평균_지상층수,
  AVG(지하_층_수) AS 평균_지하층수,
  AVG("부속_건축물_수") AS 평균_부속_건축물_수,
  AVG("부속_건축물_면적(㎡)") AS 평균_부속_건축물_면적,
  AVG("총_동_연면적(㎡)") AS 평균_총_동_연면적,
  AVG("총괄_대지_면적(㎡)") AS 평균_총괄_대지_면적,
  AVG("총괄_건축_면적(㎡)") AS 평균_총괄_건축_면적,
  AVG("총괄_건폐_율(%)") AS 평균_총괄_건폐율,
  AVG("총괄_연면적(㎡)") AS 평균_총괄_연면적,
  AVG("총괄_용적_률_산정_연면적(㎡)") AS 평균_총괄_용적률산정용연면적,
  AVG("총괄_용적_률(%)") AS 평균_총괄_용적률,
  AVG("총괄_주_건축물_수") AS 평균_총괄_주_건축물_수,
FROM
  기본개요_표제부_총괄표제부
GROUP BY
  관리_상위_건축물대장_PK_null,
  주_부속_구분_코드
ORDER BY
  관리_상위_건축물대장_PK_null DESC,
  주_부속_구분_코드;
"""
con.sql(query).show()

┌──────────────────────────────┬───────────────────┬───────────┬────────────────────┬────────────────────┬─────────────────────┬────────────────────┬─────────────────────────┬────────────────────┬────────────────────┬──────────────────────┬─────────────────────┬───────────────────────┬────────────────────┬─────────────────────┬─────────────────────┬────────────────────┬───────────────────┬──────────────────────────────┬────────────────────┬────────────────────────┐
│ 관리_상위_건축물대장_PK_null │ 주_부속_구분_코드 │ 건축물_수 │   평균_대지_면적   │   평균_건축_면적   │     평균_건폐율     │    평균_연면적     │ 평균_용적률산정용연면적 │    평균_용적률     │   평균_지상층수    │    평균_지하층수     │ 평균_부속_건축물_수 │ 평균_부속_건축물_면적 │ 평균_총_동_연면적  │ 평균_총괄_대지_면적 │ 평균_총괄_건축_면적 │  평균_총괄_건폐율  │ 평균_총괄_연면적  │ 평균_총괄_용적률산정용연면적 │  평균_총괄_용적률  │ 평균_총괄_주_건축물_수 │
│           boolean            │      varchar      │   int64   │       double       │       double       │       double        │       double       │         double          │       double       │       double    

총괄표제부가 없는 단독 주건축물에서 평균 건폐율이 13,000%로 나타나는 등, 이상값의 영향이 매우 크다.

In [15]:
query = """
SELECT * FROM 기본개요_표제부_총괄표제부
WHERE
  주_부속_구분_코드 = 0
"""
주건축물 = con.sql(query)
주건축물.limit(5).show()

┌────────────────────┬─────────────────────────┬──────────────────────────────┬──────────────────────────────────┬──────────────┬─────────────────┬─────────────┬─────────────┬───────────────────┬──────────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬──────────────┬─────────────────┬─────────────────────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬──────────┬──────────┬─────────────┬───────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬──────────────────────┬────────────────────────┬────────────────────┬────────────────────┬───────────────────┬─────────────────────┬───────────────────────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │ 관리_상위_건축물대장_PK_null │            대지_위치             │   건물_명    │      동_명      │ 시군구_코드 │ 법정동_코드 │ 주_부속_구분_코드 │ 주_부속_구분_코드_명 │ 

In [16]:
con.sql("SELECT COUNT(*) FROM 주건축물").show()

┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│      7364967 │
└──────────────┘



건축물대장 표제부 중, 기본개요와 연계 성공하고, 주/부속 구분 코드가 0(주건축물)으로 기재된 경우는 7,364,967 건으로, 2024년 건축물 현황 통계 7,421,603동과 비교할 때 다소 작은 수치이다.

이러한 차이는 보안시설 등 일부 건축물 데이터가 개방데이터에서 제외된 것이 주 원인으로 판단되나, 기본개요 연계 실패, 주/부속 구분 코드 기재 오류 등의 경우도 배제할 수 없다.

In [17]:
query = """
SELECT * FROM 주건축물
WHERE
  관리_상위_건축물대장_PK_null IS true
"""
주건축물_단독 = con.sql(query)
주건축물_단독.limit(5).show()

┌────────────────────────┬─────────────────────────┬──────────────────────────────┬─────────────────────────────────────────┬─────────┬─────────┬─────────────┬─────────────┬───────────────────┬──────────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬──────────────┬───────────────────┬───────────────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬──────────┬──────────┬─────────────┬───────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬──────────────────────┬────────────────┬────────────────────┬────────────────────┬───────────────────┬─────────────────────┬───────────────────────────┐
│   관리_건축물대장_PK   │ 관리_상위_건축물대장_PK │ 관리_상위_건축물대장_PK_null │                대지_위치                │ 건물_명 │  동_명  │ 시군구_코드 │ 법정동_코드 │ 주_부속_구분_코드 │ 주_부속_구분_코드_명 │ 대지_면적(㎡) │ 건축_면적

In [18]:
query_columns = """
  관리_건축물대장_PK,
  관리_상위_건축물대장_PK,
  대지_위치,
  건물_명,
  동_명,
  주_부속_구분_코드,
  "대지_면적(㎡)",
  "건축_면적(㎡)",
  "건폐_율(%)",
  "연면적(㎡)",
  "용적_률_산정_연면적(㎡)",
  "용적_률(%)",
  지상_층_수,
  지하_층_수,
  "부속_건축물_수",
  "부속_건축물_면적(㎡)",
  "총_동_연면적(㎡)",
  생성_일자,
"""
query = f"""
SELECT
{query_columns}
FROM
  주건축물_단독
ORDER BY
  RANDOM();
"""
con.sql(query).limit(20).show()

┌────────────────────┬─────────────────────────┬──────────────────────────────────────────────────┬───────────────┬─────────┬───────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬───────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │                    대지_위치                     │    건물_명    │  동_명  │ 주_부속_구분_코드 │ 대지_면적(㎡) │ 건축_면적(㎡) │  건폐_율(%)   │  연면적(㎡)   │ 용적_률_산정_연면적(㎡) │  용적_률(%)   │ 지상_층_수 │ 지하_층_수 │ 부속_건축물_수 │ 부속_건축물_면적(㎡) │ 총_동_연면적(㎡) │ 생성_일자 │
│      varchar       │         varchar         │                     varchar                      │    varchar    │ varchar │      varchar      │ decimal(18,3) │ decimal(18,3) │ decimal(18,3) │ decimal(18,3) │      decimal(18,3)      │ decimal(18,3) │   int32    │   int32    │     int32      │    decimal(18,3)     │  decimal(18,3)   │  varchar  │
├────────────────────┼────────────────────────

단독 주건축물의 경우, 대지면적, 건폐율, 용적률에 오류가 있는 경우가 많다.

건축면적, 연면적, 용적률 산정용 연면적, 지상/지하 층수의 오류율이 낮다.

총 동 연면적은 연면적과 동일하거나 0(미기재)인 경우가 많다.

→ 동별 연면적 파악을 위해서는 연면적 값 사용, 검증에는 건축면적, 층수 등 사용

건축면적과 연면적이 0(미기재)인 건축물도 다수 존재. 용적률 산정 연면적만 기재됨.

사용승인일 기준 최근 10년 건축한 건축물의 경우에도 연면적이 음수인 등 오류 사례 존재함.

- 광진리 168-1번지: 신축 내역 없이 -132.16 ㎡ 개축만 등재하여 연면적이 음수로 기재됨
- 서부리 816-18번지: 증축 후 이기된 별도 대장과 별개로 감축만 등재하여 연면적 음수 기재
- 본리리 108-2번지: 2004년 신축 사용승인 후 용도변경, 대장 정비 등 있었으나, 최종적으로 건축면적, 연면적 미기재, 용적률 산정용 연면적 음수 기재. 층별개요는 정상.

여의도 면적(제방 안쪽 면적 2.9㎢ = 2,900,000㎡)의 수 배~수십 배에 달하는 연면적이 기재된 경우 다수로, 대지면적, 건축면적 등과 비교할 때 오류임이 명백함.

연면적이 지나치게 큰 오류가 있는 경우 총 동 연면적도 같은 값이어서 오류 정정에 도움이 되지 않음.



집합건축물대장 표제부의 경우 대지면적, 건폐율, 용적률이 기재되지 않음. 또한, 해당 수치는 대지 단위로 산정되므로 동별 연면적 검증에는 도움이 되지 않음.

건축면적이 미기재된 사례도 있음.

- 한솔동 939번지 상가동: 건축면적 미기재(대장상 빈 칸으로 나타남).
  실제로는 상가동이 독립적으로 존재하는 것이 아니라 101동 하부에 위치하고 있으며,
  주동과 그 사이 공간에 위치한 1층 건물로 옥상을 공원화하여 이용하고 있음.
  실제로 1층 부분이 건축면적에 산입되지 않아 건축면적이 0이 된 것인지는 알 수 없으나,
  건축물대장에서는 빈 칸으로 나타나고 있으므로, 미기재로 판단함.

지상층수가 0으로 기재된 사례가 9,097건에 달함.
다만, 지하도상가 등 실제로 지상층이 존재하지 않는 경우도 포함된 수치임.

지상층수는 0이 반드시 미기재를 뜻하지는 않으므로 주의가 필요함.

## 요약

연면적, 용적률 산정 연면적, 대지면적, 건축면적, 건폐율, 용적률 등 데이터는 미기재를 0으로 저장하고 있음.

이외에도 명백한 오류 사례가 많음.

수치 데이터 미기재(0) 및 오류 사례를 널(null)값으로 처리하고, 검증 규칙을 미기재 및 오류를 고려하여 엄밀하게 정의할 필요가 있음.

---

In [19]:
주건축물.columns

['관리_건축물대장_PK',
 '관리_상위_건축물대장_PK',
 '관리_상위_건축물대장_PK_null',
 '대지_위치',
 '건물_명',
 '동_명',
 '시군구_코드',
 '법정동_코드',
 '주_부속_구분_코드',
 '주_부속_구분_코드_명',
 '대지_면적(㎡)',
 '건축_면적(㎡)',
 '건폐_율(%)',
 '연면적(㎡)',
 '용적_률_산정_연면적(㎡)',
 '용적_률(%)',
 '주_용도_코드',
 '주_용도_코드_명',
 '기타_용도',
 '높이(m)',
 '지상_층_수',
 '지하_층_수',
 '부속_건축물_수',
 '부속_건축물_면적(㎡)',
 '총_동_연면적(㎡)',
 '허가_일',
 '착공_일',
 '사용승인_일',
 '생성_일자',
 '총괄_대지_면적(㎡)',
 '총괄_건축_면적(㎡)',
 '총괄_건폐_율(%)',
 '총괄_연면적(㎡)',
 '총괄_용적_률_산정_연면적(㎡)',
 '총괄_용적_률(%)',
 '총괄_주_용도_코드',
 '총괄_주_용도_코드_명',
 '총괄_기타_용도',
 '총괄_세대_수(세대)',
 '총괄_가구_수(가구)',
 '총괄_주_건축물_수',
 '총괄_부속_건축물_수',
 '총괄_부속_건축물_면적(㎡)']

In [20]:
주건축물.limit(1).show()

┌────────────────────┬─────────────────────────┬──────────────────────────────┬──────────────────────────────────┬──────────┬─────────┬─────────────┬─────────────┬───────────────────┬──────────────────────┬───────────────┬───────────────┬───────────────┬───────────────┬─────────────────────────┬───────────────┬──────────────┬─────────────────┬────────────────────────┬───────────────┬────────────┬────────────┬────────────────┬──────────────────────┬──────────────────┬─────────┬─────────┬─────────────┬───────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬──────────────────────┬────────────────────────────────────────────┬────────────────────┬────────────────────┬───────────────────┬─────────────────────┬───────────────────────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │ 관리_상위_건축물대장_PK_null │            대지_위치             │ 건물_명  │  동_명  │ 시군구_코드 │ 법정동_코드 │ 주_부속_구분_코드 │ 주_부속_구분_코드_명 │ 대지_면적(㎡

In [21]:
# 0 as null

query = """
SELECT
  "관리_건축물대장_PK",
  "관리_상위_건축물대장_PK",
  "관리_상위_건축물대장_PK_null" AS 단독,
  "대지_위치",
  "시군구_코드",
  "법정동_코드",
  "주_용도_코드",
  "주_용도_코드_명",
  "기타_용도",
  CAST(NULLIF("대지_면적(㎡)", 0) AS DOUBLE) AS "대지_면적(㎡)",
  CAST(NULLIF("건축_면적(㎡)", 0) AS DOUBLE) AS "건축_면적(㎡)",
  CAST(NULLIF("건폐_율(%)", 0) AS DOUBLE) AS "건폐_율(%)",
  CAST(NULLIF("연면적(㎡)", 0) AS DOUBLE) AS "연면적(㎡)",
  CAST(NULLIF("용적_률_산정_연면적(㎡)", 0) AS DOUBLE) AS "용적_률_산정_연면적(㎡)",
  CAST(NULLIF("용적_률(%)", 0) AS DOUBLE) AS "용적_률(%)",
  CAST(NULLIF("높이(m)", 0) AS DOUBLE) AS "높이(m)",
  "지상_층_수",
  "지하_층_수",
  "사용승인_일",
  CAST(NULLIF("총괄_대지_면적(㎡)", 0) AS DOUBLE) AS "총괄_대지_면적(㎡)",
  CAST(NULLIF("총괄_건축_면적(㎡)", 0) AS DOUBLE) AS "총괄_건축_면적(㎡)",
  CAST(NULLIF("총괄_건폐_율(%)", 0) AS DOUBLE) AS "총괄_건폐_율(%)",
  CAST(NULLIF("총괄_연면적(㎡)", 0) AS DOUBLE) AS "총괄_연면적(㎡)",
  CAST(NULLIF("총괄_용적_률_산정_연면적(㎡)", 0) AS DOUBLE) AS "총괄_용적_률_산정_연면적(㎡)",
  CAST(NULLIF("총괄_용적_률(%)", 0) AS DOUBLE) AS "총괄_용적_률(%)",
  NULLIF("총괄_주_건축물_수", 0) AS "총괄_주_건축물_수",
FROM
  주건축물
"""
rel = con.sql(query)
rel.limit(20).show()
rel.to_parquet("../data/processed/주건축물_표제부_기본개요_총괄표제부.parquet")

┌────────────────────┬─────────────────────────┬─────────┬─────────────────────────────────────┬─────────────┬─────────────┬──────────────┬───────────────────┬────────────────────────┬───────────────┬───────────────┬────────────┬────────────┬─────────────────────────┬────────────┬─────────┬────────────┬────────────┬─────────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │  단독   │              대지_위치              │ 시군구_코드 │ 법정동_코드 │ 주_용도_코드 │  주_용도_코드_명  │       기타_용도        │ 대지_면적(㎡) │ 건축_면적(㎡) │ 건폐_율(%) │ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 용적_률(%) │ 높이(m) │ 지상_층_수 │ 지하_층_수 │ 사용승인_일 │ 총괄_대지_면적(㎡) │ 총괄_건축_면적(㎡) │ 총괄_건폐_율(%) │ 총괄_연면적(㎡) │ 총괄_용적_률_산정_연면적(㎡) │ 총괄_용적_률(%) │ 총괄_주_건축물_수 │
│      varchar       │         varchar         │ boolean │               varchar               │   varchar   │   varchar   │   varchar    │      varchar      │        varchar       

---

In [22]:
rel_orig = con.from_parquet(
    "../data/processed/주건축물_표제부_기본개요_총괄표제부.parquet"
)

con.sql("DESCRIBE rel_orig").show(max_rows=100)

┌──────────────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│         column_name          │ column_type │  null   │   key   │ default │  extra  │
│           varchar            │   varchar   │ varchar │ varchar │ varchar │ varchar │
├──────────────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ 관리_건축물대장_PK           │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ 관리_상위_건축물대장_PK      │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ 단독                         │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ 대지_위치                    │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ 시군구_코드                  │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ 법정동_코드                  │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ 주_용도_코드                 │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ 주_용도_코드_명              │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │

## 기존 규칙

### 14. (대지면적) 일반건축물대장 및 표제부 내 대지면적의 값이 건축면적보다 작은 경우


In [23]:
query = """
SELECT
  "대지_면적(㎡)",
  "총괄_대지_면적(㎡)",
  "건축_면적(㎡)",
  COALESCE("대지_면적(㎡)", "총괄_대지_면적(㎡)") < "건축_면적(㎡)" - 1 AS 기존14,
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌───────────────┬────────────────────┬───────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 기존14  │
│    double     │       double       │    double     │ boolean │
├───────────────┼────────────────────┼───────────────┼─────────┤
│          NULL │               NULL │          NULL │ NULL    │
│         513.0 │               NULL │        227.82 │ false   │
│          NULL │               NULL │          NULL │ NULL    │
│          NULL │               NULL │          NULL │ NULL    │
│          NULL │               NULL │          NULL │ NULL    │
└───────────────┴────────────────────┴───────────────┴─────────┘



In [24]:
query = """
SELECT *
FROM
  rel
WHERE
  기존14 = true
"""
con.sql(query).show()

┌───────────────┬────────────────────┬───────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 기존14  │
│    double     │       double       │    double     │ boolean │
├───────────────┼────────────────────┼───────────────┼─────────┤
│          52.9 │               NULL │         85.16 │ true    │
│         297.2 │               NULL │         347.0 │ true    │
│         218.2 │               NULL │         285.5 │ true    │
│          26.8 │               NULL │         42.94 │ true    │
│         476.0 │               NULL │        477.65 │ true    │
│          76.1 │               NULL │         82.45 │ true    │
│          72.7 │               NULL │         76.79 │ true    │
│          85.0 │               NULL │         91.07 │ true    │
│        3435.4 │               NULL │       4418.74 │ true    │
│          86.0 │               NULL │         89.12 │ true    │
│            ·  │                 ·  │            ·  │  ·      │
│            ·  │                 ·  │      

In [25]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 기존14, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 기존14  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 5190727 │
│ true    │    9773 │
│ NULL    │ 2164467 │
└─────────┴─────────┘



### 16. (건축면적) 일반건축물대장 및 표제부의 연면적이 표제부의 건축면적보다 작은 경우

In [26]:
query = """
SELECT
  "건축_면적(㎡)",
  "연면적(㎡)",
  "연면적(㎡)" < "건축_면적(㎡)" - 1 AS 기존16,
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌───────────────┬────────────┬─────────┐
│ 건축_면적(㎡) │ 연면적(㎡) │ 기존16  │
│    double     │   double   │ boolean │
├───────────────┼────────────┼─────────┤
│          NULL │      29.75 │ NULL    │
│        227.82 │    1659.98 │ false   │
│          NULL │     169.38 │ NULL    │
│          NULL │      44.33 │ NULL    │
│          NULL │      36.26 │ NULL    │
└───────────────┴────────────┴─────────┘



In [27]:
query = """
SELECT *
FROM
  rel
WHERE
  기존16 = true
"""
con.sql(query).show()

┌───────────────┬────────────┬─────────┐
│ 건축_면적(㎡) │ 연면적(㎡) │ 기존16  │
│    double     │   double   │ boolean │
├───────────────┼────────────┼─────────┤
│         92.33 │      87.87 │ true    │
│       4832.03 │     450.04 │ true    │
│       4832.03 │    1637.28 │ true    │
│       2130.67 │    1299.78 │ true    │
│       2130.67 │     637.08 │ true    │
│       2130.67 │     1280.0 │ true    │
│       4458.44 │    2580.32 │ true    │
│       4458.44 │     1460.8 │ true    │
│       7470.43 │     3353.2 │ true    │
│       7470.43 │     2240.0 │ true    │
│           ·   │        ·   │  ·      │
│           ·   │        ·   │  ·      │
│           ·   │        ·   │  ·      │
│        230.65 │     183.92 │ true    │
│      1233.298 │    398.409 │ true    │
│      1243.765 │     382.01 │ true    │
│      3821.645 │   2523.922 │ true    │
│        350.04 │     348.89 │ true    │
│       226.628 │     223.48 │ true    │
│         392.7 │      376.2 │ true    │
│         49.15 │      41.0

In [28]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 기존16, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 기존16  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 6647554 │
│ true    │  373505 │
│ NULL    │  343908 │
└─────────┴─────────┘



### 19. (건폐율) 일반건축물대장 및 표제부 건폐율 계산의 정확성

→ 소수점 이하 처리에 대한 명확한 법규 처리 없어 반올림 자리수가 불분명함(소숫점 이하 절사처리)


In [29]:
query = """
SELECT
  "대지_면적(㎡)",
  "건축_면적(㎡)",
  "건폐_율(%)",
  "건축_면적(㎡)" / "대지_면적(㎡)" * 100 AS 건폐율_재계산,
  ("건폐_율(%)" - 건폐율_재계산) AS 건폐율_차이,
  ABS(건폐율_차이) > 0.1 AS 기존19
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌───────────────┬───────────────┬────────────┬────────────────────┬───────────────────────┬─────────┐
│ 대지_면적(㎡) │ 건축_면적(㎡) │ 건폐_율(%) │   건폐율_재계산    │      건폐율_차이      │ 기존19  │
│    double     │    double     │   double   │       double       │        double         │ boolean │
├───────────────┼───────────────┼────────────┼────────────────────┼───────────────────────┼─────────┤
│          NULL │          NULL │       NULL │               NULL │                  NULL │ NULL    │
│         513.0 │        227.82 │       44.4 │ 44.409356725146196 │ -0.009356725146197675 │ false   │
│          NULL │          NULL │       NULL │               NULL │                  NULL │ NULL    │
│          NULL │          NULL │       NULL │               NULL │                  NULL │ NULL    │
│          NULL │          NULL │       NULL │               NULL │                  NULL │ NULL    │
└───────────────┴───────────────┴────────────┴────────────────────┴───────────────────────┴─────────┘



In [30]:
query = """
SELECT *
FROM
  rel
WHERE
  건폐율_차이 is not null
"""
con.sql(query).limit(20).show()

┌───────────────┬───────────────┬────────────┬────────────────────┬─────────────────────────┬─────────┐
│ 대지_면적(㎡) │ 건축_면적(㎡) │ 건폐_율(%) │   건폐율_재계산    │       건폐율_차이       │ 기존19  │
│    double     │    double     │   double   │       double       │         double          │ boolean │
├───────────────┼───────────────┼────────────┼────────────────────┼─────────────────────────┼─────────┤
│         513.0 │        227.82 │       44.4 │ 44.409356725146196 │   -0.009356725146197675 │ false   │
│        614.52 │        338.25 │      55.04 │  55.04296035930483 │  -0.0029603593048292964 │ false   │
│         61.88 │         45.37 │      73.29 │  73.31932773109243 │   -0.029327731092422482 │ false   │
│         152.4 │         87.72 │      57.56 │  57.55905511811024 │     0.00094488188976527 │ false   │
│         546.0 │        163.26 │     29.901 │ 29.901098901098898 │   -9.89010988980965e-05 │ false   │
│       12158.4 │        767.57 │      6.313 │  6.313083958415581 │  -8.395841558161976e-0

In [31]:
query = """
SELECT *
FROM
  rel
WHERE
  기존19 = true
"""
con.sql(query).limit(20).show()

┌───────────────┬───────────────┬────────────┬────────────────────┬──────────────────────┬─────────┐
│ 대지_면적(㎡) │ 건축_면적(㎡) │ 건폐_율(%) │   건폐율_재계산    │     건폐율_차이      │ 기존19  │
│    double     │    double     │   double   │       double       │        double        │ boolean │
├───────────────┼───────────────┼────────────┼────────────────────┼──────────────────────┼─────────┤
│        3096.4 │         60.21 │      39.21 │  1.944516212375662 │    37.26548378762434 │ true    │
│       78580.0 │       2175.43 │      18.97 │  2.768427080682107 │    16.20157291931789 │ true    │
│       14549.8 │        804.74 │      24.48 │ 5.5309351331289776 │   18.949064866871023 │ true    │
│       14549.8 │        351.89 │      24.48 │ 2.4185212167864854 │   22.061478783213516 │ true    │
│        4125.0 │        409.66 │      29.97 │  9.931151515151516 │   20.038848484848483 │ true    │
│        4125.0 │        410.53 │      29.97 │  9.952242424242423 │   20.017757575757578 │ true    │
│          83.6 

In [32]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 기존19, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 기존19  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 3942599 │
│ true    │   85541 │
│ NULL    │ 3336827 │
└─────────┴─────────┘



### 20. (용적률) 일반건축물대장 및 표제부 용적율 계산의 정확성


In [33]:
query = """
SELECT
  "대지_면적(㎡)",
  "용적_률_산정_연면적(㎡)",
  "용적_률(%)",
  "용적_률_산정_연면적(㎡)" / "대지_면적(㎡)" * 100 AS 용적률_재계산,
  ("용적_률(%)" - 용적률_재계산) AS 용적률_차이,
  ABS(용적률_차이) > 0.1 AS 기존20
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌───────────────┬─────────────────────────┬────────────┬────────────────────┬────────────────────┬─────────┐
│ 대지_면적(㎡) │ 용적_률_산정_연면적(㎡) │ 용적_률(%) │   용적률_재계산    │    용적률_차이     │ 기존20  │
│    double     │         double          │   double   │       double       │       double       │ boolean │
├───────────────┼─────────────────────────┼────────────┼────────────────────┼────────────────────┼─────────┤
│          NULL │                    NULL │       NULL │               NULL │               NULL │ NULL    │
│         513.0 │                  659.85 │     281.04 │ 128.62573099415204 │ 152.41426900584798 │ true    │
│          NULL │                    NULL │       NULL │               NULL │               NULL │ NULL    │
│          NULL │                    NULL │       NULL │               NULL │               NULL │ NULL    │
│          NULL │                    NULL │       NULL │               NULL │               NULL │ NULL    │
└───────────────┴─────────────────────────┴──────

In [34]:
query = """
SELECT *
FROM
  rel
WHERE
  용적률_차이 is not null
"""
con.sql(query).limit(20).show()

┌───────────────┬─────────────────────────┬────────────┬────────────────────┬────────────────────────┬─────────┐
│ 대지_면적(㎡) │ 용적_률_산정_연면적(㎡) │ 용적_률(%) │   용적률_재계산    │      용적률_차이       │ 기존20  │
│    double     │         double          │   double   │       double       │         double         │ boolean │
├───────────────┼─────────────────────────┼────────────┼────────────────────┼────────────────────────┼─────────┤
│         513.0 │                  659.85 │     281.04 │ 128.62573099415204 │     152.41426900584798 │ true    │
│        614.52 │                 1687.41 │     274.59 │ 274.58992384299944 │  7.615700053520413e-05 │ false   │
│         546.0 │                   265.6 │      48.64 │  48.64468864468865 │  -0.004688644688648935 │ false   │
│       12158.4 │                 6026.26 │     49.565 │ 49.564580865903416 │  0.0004191340965817858 │ false   │
│        1266.0 │                    56.7 │       4.48 │  4.478672985781991 │  0.0013270142180097366 │ false   │
│         540

In [35]:
query = """
SELECT *
FROM
  rel
WHERE
  기존20 = true
"""
con.sql(query).limit(20).show()

┌───────────────┬─────────────────────────┬────────────┬────────────────────┬──────────────────────┬─────────┐
│ 대지_면적(㎡) │ 용적_률_산정_연면적(㎡) │ 용적_률(%) │   용적률_재계산    │     용적률_차이      │ 기존20  │
│    double     │         double          │   double   │       double       │        double        │ boolean │
├───────────────┼─────────────────────────┼────────────┼────────────────────┼──────────────────────┼─────────┤
│         513.0 │                  659.85 │     281.04 │ 128.62573099415204 │   152.41426900584798 │ true    │
│       78580.0 │                10325.77 │      69.57 │ 13.140455586663274 │    56.42954441333672 │ true    │
│       14549.8 │                13107.33 │     335.56 │   90.0859805633067 │    245.4740194366933 │ true    │
│       14549.8 │                  891.79 │     335.56 │  6.129225143988234 │    329.4307748560118 │ true    │
│         368.3 │                  154.52 │      59.78 │ 41.954928047787135 │   17.825071952212866 │ true    │
│          40.8 │              

In [36]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 기존20, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 기존20  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 3886770 │
│ true    │   82136 │
│ NULL    │ 3396061 │
└─────────┴─────────┘



## 추가 규칙

### 01. 연면적이 용적률 산정 연면적보다 작음

In [37]:
query = """
SELECT
  "연면적(㎡)",
  "용적_률_산정_연면적(㎡)",
  "연면적(㎡)" < "용적_률_산정_연면적(㎡)" - 1 AS 신규01
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌────────────┬─────────────────────────┬─────────┐
│ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 신규01  │
│   double   │         double          │ boolean │
├────────────┼─────────────────────────┼─────────┤
│      29.75 │                    NULL │ NULL    │
│    1659.98 │                  659.85 │ false   │
│     169.38 │                    NULL │ NULL    │
│      44.33 │                    NULL │ NULL    │
│      36.26 │                    NULL │ NULL    │
└────────────┴─────────────────────────┴─────────┘



In [38]:
query = """
SELECT *
FROM
  rel
WHERE
  신규01 is not null
"""
con.sql(query).limit(20).show()

┌────────────┬─────────────────────────┬─────────┐
│ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 신규01  │
│   double   │         double          │ boolean │
├────────────┼─────────────────────────┼─────────┤
│    1659.98 │                  659.85 │ false   │
│     314.84 │                  236.47 │ false   │
│     693.41 │                  531.52 │ false   │
│      267.8 │                  184.58 │ false   │
│     232.85 │                  198.81 │ false   │
│     126.96 │                  126.96 │ false   │
│     380.18 │                   201.4 │ false   │
│    1147.64 │                 1147.64 │ false   │
│    2025.42 │                 1687.41 │ false   │
│     120.18 │                  120.18 │ false   │
│      114.9 │                   57.45 │ false   │
│     492.72 │                  492.72 │ false   │
│    2399.01 │                 2298.66 │ false   │
│    2009.93 │                 2009.93 │ false   │
│   12132.07 │                11797.99 │ false   │
│      853.7 │                  682.96 │ false

In [39]:
query = """
SELECT *
FROM
  rel
WHERE
  신규01 = true
"""
con.sql(query).limit(20).show()

┌────────────┬─────────────────────────┬─────────┐
│ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 신규01  │
│   double   │         double          │ boolean │
├────────────┼─────────────────────────┼─────────┤
│      87.87 │                   92.33 │ true    │
│    9803.01 │                10609.94 │ true    │
│     225.02 │                 2549.17 │ true    │
│     245.73 │                 1012.04 │ true    │
│      84.88 │                  207.03 │ true    │
│     149.54 │                 4224.22 │ true    │
│    1747.75 │                 1888.74 │ true    │
│      81.57 │                 2549.17 │ true    │
│     152.95 │                  179.84 │ true    │
│     347.88 │                 1012.04 │ true    │
│      71.66 │                  179.84 │ true    │
│      269.8 │                 4224.22 │ true    │
│      70.31 │                  179.84 │ true    │
│     284.82 │                  340.86 │ true    │
│     731.12 │                  733.82 │ true    │
│       88.0 │                 2549.17 │ true 

In [40]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 신규01, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 신규01  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 6999231 │
│ true    │   19295 │
│ NULL    │  346441 │
└─────────┴─────────┘



### 02. 연면적이 (건축면적×지상층수)+(대지면적×지하층수)보다 큼

In [41]:
query = """
SELECT
  "대지_면적(㎡)",
  "총괄_대지_면적(㎡)",
  "건축_면적(㎡)",
  "연면적(㎡)",
  지상_층_수,
  지하_층_수,
  ("건축_면적(㎡)" * 지상_층_수) + (COALESCE("대지_면적(㎡)", "총괄_대지_면적(㎡)") * 지하_층_수) AS 연면적_상한,
  "연면적(㎡)" > 연면적_상한 + 1 AS 신규02
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌───────────────┬────────────────────┬───────────────┬────────────┬────────────┬────────────┬─────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 연면적(㎡) │ 지상_층_수 │ 지하_층_수 │ 연면적_상한 │ 신규02  │
│    double     │       double       │    double     │   double   │   int32    │   int32    │   double    │ boolean │
├───────────────┼────────────────────┼───────────────┼────────────┼────────────┼────────────┼─────────────┼─────────┤
│          NULL │               NULL │          NULL │      29.75 │          1 │          0 │        NULL │ NULL    │
│         513.0 │               NULL │        227.82 │    1659.98 │          7 │          1 │     2107.74 │ false   │
│          NULL │               NULL │          NULL │     169.38 │          2 │          0 │        NULL │ NULL    │
│          NULL │               NULL │          NULL │      44.33 │          1 │          0 │        NULL │ NULL    │
│          NULL │               NULL │          NULL │      36.26 │          1 │          0 

In [42]:
query = """
SELECT *
FROM
  rel
WHERE
  신규02 is not null
"""
con.sql(query).limit(20).show()

┌───────────────┬────────────────────┬───────────────┬────────────┬────────────┬────────────┬────────────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 연면적(㎡) │ 지상_층_수 │ 지하_층_수 │    연면적_상한     │ 신규02  │
│    double     │       double       │    double     │   double   │   int32    │   int32    │       double       │ boolean │
├───────────────┼────────────────────┼───────────────┼────────────┼────────────┼────────────┼────────────────────┼─────────┤
│         513.0 │               NULL │        227.82 │    1659.98 │          7 │          1 │            2107.74 │ false   │
│          NULL │              658.0 │        136.18 │     314.84 │          2 │          1 │             930.36 │ false   │
│          NULL │              793.0 │        141.82 │     693.41 │          4 │          1 │            1360.28 │ false   │
│          NULL │              624.0 │        120.47 │      267.8 │          2 │          1 │             864.94 │ false   │
│          NULL │              624.0

In [43]:
query = """
SELECT *
FROM
  rel
WHERE
  신규02 = true
"""
con.sql(query).limit(20).show()

┌───────────────┬────────────────────┬───────────────┬────────────┬────────────┬────────────┬────────────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 연면적(㎡) │ 지상_층_수 │ 지하_층_수 │    연면적_상한     │ 신규02  │
│    double     │       double       │    double     │   double   │   int32    │   int32    │       double       │ boolean │
├───────────────┼────────────────────┼───────────────┼────────────┼────────────┼────────────┼────────────────────┼─────────┤
│          NULL │              331.0 │         47.54 │     120.18 │          2 │          0 │              95.08 │ true    │
│          NULL │            56242.6 │       1102.48 │    4449.59 │          4 │          0 │            4409.92 │ true    │
│        4097.1 │               NULL │         771.6 │    1307.85 │          1 │          0 │              771.6 │ true    │
│          67.4 │               NULL │         48.45 │     124.33 │          2 │          0 │               96.9 │ true    │
│          39.7 │               NULL

In [44]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 신규02, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 신규02  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 5134625 │
│ true    │   63313 │
│ NULL    │ 2167029 │
└─────────┴─────────┘



### 03. 연면적, 용적률 산정 연면적, 대지면적, 건축면적 중 음수 존재

In [48]:
query = """
SELECT
  "대지_면적(㎡)",
  "총괄_대지_면적(㎡)",
  "건축_면적(㎡)",
  "총괄_건축_면적(㎡)",
  "연면적(㎡)",
  "용적_률_산정_연면적(㎡)",
  지상_층_수,
  지하_층_수,
  COALESCE("대지_면적(㎡)", "총괄_대지_면적(㎡)", 0) < 0
  OR COALESCE("건축_면적(㎡)", "총괄_건축_면적(㎡)", 0) < 0
  OR COALESCE("연면적(㎡)", 0) < 0
  OR COALESCE("지상_층_수", 0) < 0
  OR COALESCE("지하_층_수", 0) < 0
  AS 신규03,
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(5).show()

┌───────────────┬────────────────────┬───────────────┬────────────────────┬────────────┬─────────────────────────┬────────────┬────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 총괄_건축_면적(㎡) │ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 지상_층_수 │ 지하_층_수 │ 신규03  │
│    double     │       double       │    double     │       double       │   double   │         double          │   int32    │   int32    │ boolean │
├───────────────┼────────────────────┼───────────────┼────────────────────┼────────────┼─────────────────────────┼────────────┼────────────┼─────────┤
│          NULL │               NULL │          NULL │               NULL │      29.75 │                    NULL │          1 │          0 │ false   │
│         513.0 │               NULL │        227.82 │               NULL │    1659.98 │                  659.85 │          7 │          1 │ false   │
│          NULL │               NULL │          NULL │               NULL │     169.38 │                    NULL │          2 │          0 │ 

In [50]:
query = """
SELECT *
FROM
  rel
WHERE
  신규03 = true
"""
con.sql(query).limit(5).show()

┌───────────────┬────────────────────┬───────────────┬────────────────────┬────────────┬─────────────────────────┬────────────┬────────────┬─────────┐
│ 대지_면적(㎡) │ 총괄_대지_면적(㎡) │ 건축_면적(㎡) │ 총괄_건축_면적(㎡) │ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 지상_층_수 │ 지하_층_수 │ 신규03  │
│    double     │       double       │    double     │       double       │   double   │         double          │   int32    │   int32    │ boolean │
├───────────────┼────────────────────┼───────────────┼────────────────────┼────────────┼─────────────────────────┼────────────┼────────────┼─────────┤
│         287.5 │               NULL │         200.0 │               NULL │      657.2 │                  612.96 │          4 │         -1 │ true    │
│         287.5 │               NULL │         200.0 │               NULL │     661.28 │                  617.04 │          4 │         -1 │ true    │
│          NULL │           440375.4 │       5409.27 │          142032.25 │   25167.59 │                19970.66 │          4 │         -1 │ 

In [49]:
# 컬럼의 값별 개수 집계
con.sql("""
    SELECT 신규03, COUNT(*) AS count
    FROM rel
    GROUP BY ALL
    ORDER BY ALL
""").show()

┌─────────┬─────────┐
│ 신규03  │  count  │
│ boolean │  int64  │
├─────────┼─────────┤
│ false   │ 7364947 │
│ true    │      20 │
└─────────┴─────────┘



### 결과 저장

In [51]:
# save the final result to parquet

query = """
SELECT
  *,
  COALESCE("대지_면적(㎡)", "총괄_대지_면적(㎡)") < "건축_면적(㎡)" - 1 AS 기존14,
  "연면적(㎡)" < "건축_면적(㎡)" - 1 AS 기존16,
  "건축_면적(㎡)" / "대지_면적(㎡)" * 100 AS 건폐율_재계산,
  ("건폐_율(%)" - 건폐율_재계산) AS 건폐율_차이,
  ABS(건폐율_차이) > 0.1 AS 기존19,
  "용적_률_산정_연면적(㎡)" / "대지_면적(㎡)" * 100 AS 용적률_재계산,
  ("용적_률(%)" - 용적률_재계산) AS 용적률_차이,
  ABS(용적률_차이) > 0.1 AS 기존20,
  "연면적(㎡)" < "용적_률_산정_연면적(㎡)" - 1 AS 신규01,
  ("건축_면적(㎡)" * 지상_층_수) + (COALESCE("대지_면적(㎡)", "총괄_대지_면적(㎡)") * 지하_층_수) AS 연면적_상한,
  "연면적(㎡)" > 연면적_상한 + 1 AS 신규02,
  COALESCE("대지_면적(㎡)", "총괄_대지_면적(㎡)", 0) < 0
  OR COALESCE("건축_면적(㎡)", "총괄_건축_면적(㎡)", 0) < 0
  OR COALESCE("연면적(㎡)", 0) < 0
  OR COALESCE("지상_층_수", 0) < 0
  OR COALESCE("지하_층_수", 0) < 0
  AS 신규03,
FROM
  rel_orig
"""
rel = con.sql(query)
rel.limit(20).show()
rel.to_parquet("../data/processed/주건축물_검증규칙.parquet")

┌────────────────────┬─────────────────────────┬─────────┬─────────────────────────────────────┬─────────────┬─────────────┬──────────────┬───────────────────┬──────────────────────────────────────┬───────────────┬───────────────┬────────────┬────────────┬─────────────────────────┬────────────┬─────────┬────────────┬────────────┬─────────────┬────────────────────┬────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬─────────────────┬───────────────────┬─────────┬─────────┬────────────────────┬────────────────────────┬─────────┬────────────────────┬───────────────────────┬─────────┬─────────┬─────────────┬─────────┬─────────┐
│ 관리_건축물대장_PK │ 관리_상위_건축물대장_PK │  단독   │              대지_위치              │ 시군구_코드 │ 법정동_코드 │ 주_용도_코드 │  주_용도_코드_명  │              기타_용도               │ 대지_면적(㎡) │ 건축_면적(㎡) │ 건폐_율(%) │ 연면적(㎡) │ 용적_률_산정_연면적(㎡) │ 용적_률(%) │ 높이(m) │ 지상_층_수 │ 지하_층_수 │ 사용승인_일 │ 총괄_대지_면적(㎡) │ 총괄_건축_면적(㎡) │ 총괄_건폐_율(%) │ 총괄_연면적(㎡) │ 총괄_용적_률_산정_연면적(㎡) │ 총괄_용